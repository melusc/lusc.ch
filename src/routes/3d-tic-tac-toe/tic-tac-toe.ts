import type {Opaque} from 'type-fest';
// eslint-disable-next-line n/file-extension-in-import
import {get, writable, type Readable, type Writable} from 'svelte/store';

import {TypedEventTarget} from './typed-event-target.ts';

export const enum Player {
	p1 = 1,
	p2 = 2,
}

export type Cell = {
	readonly index: number;
	readonly content: Readable<undefined | Player>;
};
type InternalCell = {
	readonly index: number;
	readonly content: Writable<undefined | Player>;
};

export type Layer = readonly Cell[];

export type Group = Opaque<readonly Cell[], 'group'>;

function * getStart(direction: number): Iterable<number> {
	if (direction === 1) {
		yield 0;
	} else if (direction === -1) {
		yield 3;
	} else {
		yield 0;
		yield 1;
		yield 2;
		yield 3;
	}
}

// x is left right
// y is front back
// z is top bottom
export function resolveIndex(x: number, y: number, z: number): number {
	// prettier-ignore
	return x + (y * 4) + (z * 16);
}

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
function createTurnStore() {
	const {subscribe, update} = writable(Player.p1);

	return {
		subscribe,
		toggle(): void {
			update(player => (player === Player.p1 ? Player.p2 : Player.p1));
		},
	};
}

export class TicTacToe extends TypedEventTarget<{
	win: {
		winner: Player;
		winningGroups: readonly Group[];
	};
}> {
	isFinished = false;
	turn = createTurnStore();

	readonly #cells: readonly InternalCell[] = Array.from(
		{length: 4 ** 3},
		(_v, index) => ({
			index,
			content: writable(undefined),
			_content: undefined,
		}),
	);

	* getGroups(): Iterable<Group> {
		yield * this.getGroupByDir({x: 1});
		yield * this.getGroupByDir({y: 1});
		yield * this.getGroupByDir({z: 1});

		yield * this.get2dDiagonals();
		yield * this.get3dDiagonals();
	}

	getLayers(): readonly Layer[] {
		const layers: Layer[] = [];
		for (let i = 0; i < 4; ++i) {
			layers.push(this.#cells.slice(i * 16, (i + 1) * 16));
		}

		return layers;
	}

	makeMove(index: number): this {
		const cells = [...this.#cells];
		const cell = cells[index];

		if (!cell) {
			throw new Error('Index out of bounds.');
		}

		if (get(cell.content) !== undefined) {
			throw new Error('Cell is already filled.');
		}

		this.#cells[index]!.content.set(get(this.turn));

		this.turn.toggle();

		this.emitOnWinner();
	}

	protected * getGroupByDir(dir: {
		x?: 1 | -1;
		y?: 1 | -1;
		z?: 1 | -1;
	}): Iterable<Group> {
		const dirX = dir.x ?? 0;
		const dirY = dir.y ?? 0;
		const dirZ = dir.z ?? 0;

		if (dirX === 0 && dirY === 0 && dirZ === 0) {
			throw new Error('Expected at least on of directions to be non-null.');
		}

		for (const x of getStart(dirX)) {
			for (const y of getStart(dirY)) {
				for (const z of getStart(dirZ)) {
					const result: Cell[] = [];
					for (let i = 0; i < 4; ++i) {
						result.push(
							this.#cells[
								// prettier-ignore
								resolveIndex(x + (dirX * i), y + (dirY * i), z + (dirZ * i))
							]!,
						);
					}

					yield [...result] as const as Group;
				}
			}
		}
	}

	// There are four diagonals, ignoring direction
	// Always go top to bottom, so z = 1
	protected * get3dDiagonals(): Iterable<Group> {
		const directions = [1, -1] as const;

		for (const x of directions) {
			for (const y of directions) {
				yield * this.getGroupByDir({x, y, z: 1});
			}
		}
	}

	protected * get2dDiagonals(): Iterable<Group> {
		yield * this.getGroupByDir({x: 1, y: 1});
		yield * this.getGroupByDir({x: -1, y: 1});

		yield * this.getGroupByDir({x: 1, z: 1});
		yield * this.getGroupByDir({x: -1, z: 1});
		yield * this.getGroupByDir({y: 1, z: 1});
		yield * this.getGroupByDir({y: -1, z: 1});
	}

	protected isGroupWinning(group: Group): false | Player {
		let lastPlayer: Player | undefined;

		for (const cell of group) {
			const content = get(cell.content);

			if (content === undefined) {
				return false;
			}

			if (lastPlayer !== undefined && lastPlayer !== content) {
				return false;
			}

			lastPlayer = content;
		}

		return lastPlayer!;
	}

	protected emitOnWinner(): void {
		if (this.isFinished) {
			throw new Error('Game is already over.');
		}

		const groups: Group[] = [];
		let winningPlayer: Player | undefined;
		for (const group of this.getGroups()) {
			const groupWinner = this.isGroupWinning(group);
			if (groupWinner !== false) {
				if (winningPlayer !== undefined && winningPlayer !== groupWinner) {
					// This should be impossible, because it doesn't get here if `isFinished === true`
					throw new Error('Got two winners.');
				}

				winningPlayer = groupWinner;
				groups.push(group);
			}
		}

		if (winningPlayer === undefined) {
			return;
		}

		this.isFinished = true;

		this.emit('win', {
			winner: winningPlayer,
			winningGroups: groups,
		});
	}
}
