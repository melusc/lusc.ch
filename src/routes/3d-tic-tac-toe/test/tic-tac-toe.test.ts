import {expect, test, vi} from 'vitest';

import {TicTacToe, resolveIndex} from '../tic-tac-toe.ts';

test('#resolveIndex', () => {
	let i = 0;

	for (let z = 0; z < 4; ++z) {
		for (let y = 0; y < 4; ++y) {
			for (let x = 0; x < 4; ++x) {
				expect(resolveIndex(x, y, z)).toStrictEqual(i++);
			}
		}
	}
});

test('#getGroups', () => {
	const t = new TicTacToe();

	const foundGroups = new Set<string>();

	for (const group of t.getGroups()) {
		foundGroups.add(group.map(({index}) => index).join(', '));
	}

	// prettier-ignore
	const expectedSize
		= (3 * 4 * 4) // "Flat" groups, like rows, 3 directions, 4 layers, 4 per layer
		+ (2 * 4 * 3) // 2d diagonals, 2 per layer, 4 layers, 3 directiosn
		+ 4; // 3d diagonals

	expect(foundGroups.size).toStrictEqual(expectedSize);

	expect(foundGroups).toMatchSnapshot();
});

test('Winning', () => {
	const t = new TicTacToe();

	const handler = vi.fn();
	t.on('win', handler);

	for (const move of [0, 4, 1, 5, 2, 6, 3]) {
		t.makeMove(move);
	}

	expect(handler).toHaveBeenCalledOnce();
});
