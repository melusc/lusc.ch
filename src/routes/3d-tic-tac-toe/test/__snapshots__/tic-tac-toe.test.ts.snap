// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#getGroups 1`] = `
Set {
  "0, 1, 2, 3",
  "0, 16, 32, 48",
  "0, 17, 34, 51",
  "0, 20, 40, 60",
  "0, 21, 42, 63",
  "0, 4, 8, 12",
  "0, 5, 10, 15",
  "1, 17, 33, 49",
  "1, 21, 41, 61",
  "1, 5, 9, 13",
  "10, 26, 42, 58",
  "11, 26, 41, 56",
  "11, 27, 43, 59",
  "12, 13, 14, 15",
  "12, 24, 36, 48",
  "12, 25, 38, 51",
  "12, 28, 44, 60",
  "12, 29, 46, 63",
  "13, 25, 37, 49",
  "13, 29, 45, 61",
  "14, 26, 38, 50",
  "14, 30, 46, 62",
  "15, 26, 37, 48",
  "15, 27, 39, 51",
  "15, 30, 45, 60",
  "15, 31, 47, 63",
  "16, 17, 18, 19",
  "16, 20, 24, 28",
  "16, 21, 26, 31",
  "17, 21, 25, 29",
  "18, 22, 26, 30",
  "19, 22, 25, 28",
  "19, 23, 27, 31",
  "2, 18, 34, 50",
  "2, 22, 42, 62",
  "2, 6, 10, 14",
  "20, 21, 22, 23",
  "24, 25, 26, 27",
  "28, 29, 30, 31",
  "3, 18, 33, 48",
  "3, 19, 35, 51",
  "3, 22, 41, 60",
  "3, 23, 43, 63",
  "3, 6, 9, 12",
  "3, 7, 11, 15",
  "32, 33, 34, 35",
  "32, 36, 40, 44",
  "32, 37, 42, 47",
  "33, 37, 41, 45",
  "34, 38, 42, 46",
  "35, 38, 41, 44",
  "35, 39, 43, 47",
  "36, 37, 38, 39",
  "4, 20, 36, 52",
  "4, 21, 38, 55",
  "4, 5, 6, 7",
  "40, 41, 42, 43",
  "44, 45, 46, 47",
  "48, 49, 50, 51",
  "48, 52, 56, 60",
  "48, 53, 58, 63",
  "49, 53, 57, 61",
  "5, 21, 37, 53",
  "50, 54, 58, 62",
  "51, 54, 57, 60",
  "51, 55, 59, 63",
  "52, 53, 54, 55",
  "56, 57, 58, 59",
  "6, 22, 38, 54",
  "60, 61, 62, 63",
  "7, 22, 37, 52",
  "7, 23, 39, 55",
  "8, 24, 40, 56",
  "8, 25, 42, 59",
  "8, 9, 10, 11",
  "9, 25, 41, 57",
}
`;
